### ISSUE TITLE
Segmentation fault when Smarty::fetch() is called on a Smarty_Internal_Template object

### ISSUE THREAD

**Issue by arkonan** [tideways/php-profiler-extension#44]
From the smarty code documentation for `Smarty_Internal_TemplateBase::fetch()` the `$template` parameter can be `null`, a string or a template object:

```php
    /**
     * fetches a rendered Smarty template
     *
     * @param  string $template   the resource handle of the template file or template object
     * @param  mixed  $cache_id   cache id to be used with this template
     * @param  mixed  $compile_id compile id to be used with this template
     * @param  object $parent     next higher level of Smarty variables
     *
     * @throws Exception
     * @throws SmartyException
     * @return string rendered template output
     */
    public function fetch($template = null, $cache_id = null, $compile_id = null, $parent = null)
```

`tw_trace_callback_smarty3_template()` does not account for passing a Smarty_Internal_Template object as the `$template` parameter.  If I'm reading it correctly it assumes that either `$template` is a string, or `fetch()` was called on a Smarty_Internal_Template. 

Works:
```php
$smarty->fetch($template_string);
```

Also works:
```php
$template_object = $smarty->createTemplate($template_string, $cache_id, $compile_id, $smarty);
$template_object->fetch();
```

Crashes:
```php
$template_object = $smarty->createTemplate($template_string, $cache_id, $compile_id, $smarty);
$smarty->fetch($template_object);
```

Crash trace:
```
Process:               php-cgi [97896]
Path:                  /Applications/MAMP/*/php-cgi
Identifier:            php-cgi
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        php-cgi [97368]
Responsible:           php-cgi [97896]
User ID:               502

Date/Time:             2017-12-05 16:56:49.107 -0500
OS Version:            Mac OS X 10.13.1 (17B1003)
Report Version:        12
Anonymous UUID:        B341CCF6-5D1C-42CB-4AE5-CB93D10D7716


Time Awake Since Boot: 110000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [0]

VM Regions Near 0x8:
--> 
    __TEXT                 000000010e57d000-000000010f541000 [ 15.8M] r-x/rwx SM=COW  /Applications/MAMP/*

Application Specific Information:
crashed on child side of fork pre-exec

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   php-cgi                         0x000000010ee91e20 zend_std_read_property + 3136
1   php-cgi                         0x000000010ee19276 zend_read_property + 646
2   tideways.so                     0x0000000112c11cdc tw_trace_callback_smarty3_template + 204
3   tideways.so                     0x0000000112c1706f hp_mode_hier_beginfn_cb + 95
4   tideways.so                     0x0000000112c0fc37 hp_execute_ex + 1079
5   php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
6   php-cgi                         0x000000010eeac034 execute_ex + 36
7   tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
8   php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
9   php-cgi                         0x000000010eeac034 execute_ex + 36
10  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
11  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
12  php-cgi                         0x000000010eeac034 execute_ex + 36
13  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
14  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
15  php-cgi                         0x000000010eeac034 execute_ex + 36
16  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
17  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
18  php-cgi                         0x000000010eeac034 execute_ex + 36
19  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
20  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
21  php-cgi                         0x000000010eeac034 execute_ex + 36
22  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
23  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
24  php-cgi                         0x000000010eeac034 execute_ex + 36
25  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
26  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
27  php-cgi                         0x000000010eeac034 execute_ex + 36
28  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
29  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
30  php-cgi                         0x000000010eeac034 execute_ex + 36
31  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
32  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
33  php-cgi                         0x000000010eeac034 execute_ex + 36
34  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
35  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
36  php-cgi                         0x000000010eeac034 execute_ex + 36
37  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
38  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
39  php-cgi                         0x000000010eeac034 execute_ex + 36
40  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
41  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
42  php-cgi                         0x000000010eeac034 execute_ex + 36
43  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
44  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
45  php-cgi                         0x000000010eeac034 execute_ex + 36
46  tideways.so                     0x0000000112c0fc49 hp_execute_ex + 1097
47  php-cgi                         0x000000010eef35b3 ZEND_DO_FCALL_SPEC_HANDLER + 3523
48  php-cgi                         0x000000010eeac034 execute_ex + 36
49  php-cgi                         0x000000010eeacb46 zend_execute + 2710
50  php-cgi                         0x000000010ee00347 zend_execute_scripts + 615
51  php-cgi                         0x000000010ed2426f php_execute_script + 1615
52  php-cgi                         0x000000010efea53c main + 8940
53  libdyld.dylib                   0x00007fff64542145 start + 1
```

Let me know if you need any additional information.  In the meantime I have modified my code to call `fetch()` on the template object.

**Comment by beberlei** [comment]
@arkonan I will look into it immediately, thank you for the report.

**Comment by beberlei** [comment]
Fixed in https://github.com/tideways/php-profiler-extension/commit/8842bb37bffe775480d47205347a9c6a26181b05 - can you test by compiling yourself on MacOS? Otherwise I can notify you again when I finished uploading the v4.1.5 to the homebrew repository.

**Comment by arkonan** [comment]
Confirmed fixed after 8842bb3.  Thanks for the quick response!

**Comment by beberlei** [comment]
@arkonan thank you for the detailed report, it was very easy to find :-)
